infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -0,0 +1,210 @@
+    }
+
+    /**
+     * The cache limited by size.
+     * For a synchronous cache we are using a same-thread executor (Runnable::run)
+     * https://github.com/ben-manes/caffeine/issues/90
+     *
+     * @param cacheSize cache size
+     */
+    public SequencerServerCache(long cacheSize) {
+        this.conflictCache = Caffeine.newBuilder()
+                .maximumSize(cacheSize)
+                //Performing periodic maintenance using current thread (synchronously)
+                .executor(Runnable::run)
+                .writer(getDefaultCacheWriter())
+                .recordStats()
+                .build();
+    }
+
+    private CacheWriter<ConflictTxStream, Long> getDefaultCacheWriter() {
+        return new CacheWriter<ConflictTxStream, Long>() {
+            /**
+             * Don't do any additional actions during the write operation. Let caffeine write the record into the cache
+             * @param key a key
+             * @param value a value
+             */
+            @Override
+            public void write(@Nonnull ConflictTxStream key, @Nonnull Long value) {
+                //ignore
+            }
+
+            /**
+             * Eviction policy https://github.com/ben-manes/caffeine/wiki/Writer
+             * @param key  conflict stream
+             * @param globalAddress global address
+             * @param cause a removal cause
+             */
+            @Override
+            public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull RemovalCause cause) {
+                if (cause == RemovalCause.REPLACED) {
+                    String errMsg = String.format("Override error. Conflict key: %s, address: %s", key, globalAddress);
+                    throw new IllegalStateException(errMsg);
+                }
+
+                log.trace(
+                        "Updating maxConflictWildcard. Old = '{}', new ='{}' conflictParam = '{}'. Cause = '{}'",
+                        maxConflictWildcard, globalAddress, key, cause
+                );
+
+                if (globalAddress == null) {
+                    globalAddress = Address.NOT_FOUND;
+                }
+                maxConflictWildcard = Math.max(globalAddress, maxConflictWildcard);
+            }
+        };
+    }
+
+    /**
+     * Returns the value associated with the {@code key} in this cache,
+     * or {@code null} if there is no cached value for the {@code key}.
+     *
+     * @param conflictKey conflict stream
+     * @return global address
+     */
+    public Long getIfPresent(ConflictTxStream conflictKey) {
+        return conflictCache.getIfPresent(conflictKey);
+    }
+
+    /**
+     * Invalidate all records up to a trim mark.
+     *
+     * @param trimMark trim mark
+     */
+    public void invalidateUpTo(long trimMark) {
+        log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
+
+        long entries = 0;
+        for (ConflictTxStream key : conflictCache.asMap().keySet()) {
```

Comment:
Why not taking the value from the entry set instead of getIfpresent?

Reply:
Because we want to use the API provided by the cache. Going underlying, get underlying structure and use it, that seems no so good idea to me. We get the underlying map and use it (`asMap()` method) only because there is no different ways to iterate through cache

Reply:
fyi the `asMap()` is safe to use however you prefer and is meant as an escape hatch for your more complex flows. There is no performance penalty or quirks to using it. The `Cache` interface is to be more idiomatic for the common case and better communicates intent. You can use `entrySet` or the convenience `forEach(k, v -> ...)` methods if desired (or not, of course).

Reply:
Yep, I just try to avoid leaking abstractions. If I use some class I want to use its API. I prefer to avoid extracting underlying data structures and use them directly.